rsa/rsa_factory.go

```diff
@@ -12,60 +12,49 @@
 package rsa
 
 import (
-	"bytes"
 	"crypto/rand"
 	"crypto/rsa"
 	"crypto/sha256"
 	"crypto/x509"
-	"encoding/asn1"
-	"encoding/pem"
-	"fmt"
+	"encoding/base64"
 )
 
 // GenerateRSAKeyPair generates a RSA key pair for the provided bit size.
-// Recommended bit size is 4096
-func GenerateRSAKeyPair(bitSize int) (*pem.Block, *pem.Block, error) {
+func GenerateRSAKeyPair() (*rsa.PrivateKey, error) {
 	reader := rand.Reader
+	return rsa.GenerateKey(reader, 4096)
+}
 
-	key, err := rsa.GenerateKey(reader, bitSize)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	publicKey := key.PublicKey
-
-	asn1Bytes, err := asn1.Marshal(publicKey)
-	pemkeyPublic := &pem.Block{
-		Type:  "PUBLIC KEY",
-		Bytes: asn1Bytes,
-	}
-
-	var privateKey = &pem.Block{
-		Type:  "PRIVATE KEY",
-		Bytes: x509.MarshalPKCS1PrivateKey(key),
-	}
+// EvalHash generates a SHA256 hash as string for the provided pem block
+func EvalHash(p *rsa.PublicKey) (string, error) {
+	b := x509.MarshalPKCS1PublicKey(p)
+	h := sha256.Sum256(b)
+	b64 := base64.StdEncoding.EncodeToString(h[:])
+	return b64, nil
+}
 
-	return pemkeyPublic, privateKey, nil
+// Encode converts a rsa.PublicKey to a base64 encoded pkcs1 string
+func Encode(p *rsa.PublicKey) (string, error) {
+	b := x509.MarshalPKCS1PublicKey(p)
+	s := base64.StdEncoding.EncodeToString(b)
+	return s, nil
 }
 
```

Comment:
compared with pem format, will this cause confusions if we need to share the public key with partners?

Reply:
I am not so familiar with this part, this code is from transit engine:

```
                // Encode the RSA public key in PEM format to return over the
		// API
		derBytes, err := x509.MarshalPKIXPublicKey(keyEntry.RSAKey.Public())
		if err != nil {
			return "", errwrap.Wrapf("error marshaling RSA public key: {{err}}", err)
		}
		pemBlock := &pem.Block{
			Type:  "PUBLIC KEY",
			Bytes: derBytes,
		}
		pemBytes := pem.EncodeToMemory(pemBlock)
		if pemBytes == nil || len(pemBytes) == 0 {
			return "", fmt.Errorf("failed to PEM-encode RSA public key")
		}
		return string(pemBytes), nil              
```

Reply:
We can use pem if intended for external exchange, which wasn't a requirement.